torch/utils/data/distributed.py

```diff
@@ -76,43 +77,31 @@ def __init__(self, dataset: Dataset, num_replicas: Optional[int] = None,
         self.rank = rank
         self.epoch = 0
         self.drop_last = drop_last
+        # `type:ignore` is required because Dataset cannot provide a default __len__
+        # see NOTE in pytorch/torch/utils/data/sampler.py
+        self.dataset_size = len(self.dataset)  # type: ignore
         # If the dataset length is evenly divisible by # of replicas, then there
         # is no need to drop any data, since the dataset will be split equally.
-        if self.drop_last and len(self.dataset) % self.num_replicas != 0:  # type: ignore
+        if self.drop_last and self.dataset_size % self.num_replicas != 0:
             # Split to nearest available length that is evenly divisible.
             # This is to ensure each rank receives the same amount of data when
             # using this Sampler.
-            self.num_samples = math.ceil(
-                # `type:ignore` is required because Dataset cannot provide a default __len__
-                # see NOTE in pytorch/torch/utils/data/sampler.py
-                (len(self.dataset) - self.num_replicas) / self.num_replicas  # type: ignore
-            )
+            self.num_samples = math.ceil((self.dataset_size - self.num_replicas) / self.num_replicas)
         else:
-            self.num_samples = math.ceil(len(self.dataset) / self.num_replicas)  # type: ignore
+            self.num_samples = math.ceil(self.dataset_size / self.num_replicas)
         self.total_size = self.num_samples * self.num_replicas
         self.shuffle = shuffle
         self.seed = seed
 
     def __iter__(self) -> Iterator[T_co]:
-        if self.shuffle:
+        if not self.shuffle:
+            indices = index_utils.Range(self.dataset_size, stop=self.total_size)
+        else:
             # deterministically shuffle based on epoch and seed
             g = torch.Generator()
             g.manual_seed(self.seed + self.epoch)
-            indices = torch.randperm(len(self.dataset), generator=g).tolist()  # type: ignore
-        else:
-            indices = list(range(len(self.dataset)))  # type: ignore
-
-        if not self.drop_last:
-            # add extra samples to make it evenly divisible
-            padding_size = self.total_size - len(indices)
-            if padding_size <= len(indices):
-                indices += indices[:padding_size]
-            else:
-                indices += (indices * math.ceil(padding_size / len(indices)))[:padding_size]
-        else:
-            # remove tail of data to make it evenly divisible.
-            indices = indices[:self.total_size]
-        assert len(indices) == self.total_size
+            seed = int(torch.randint(high=self.dataset_size, size=(1,), dtype=torch.int64, generator=g).item())
+            indices = index_utils.Permutation(self.dataset_size, seed, stop=self.total_size)
```

Comment:
Can you also run a benchmark for `RandomIndexGenerator` against `torch.randperm` for smaller/medium/large index number.
You can use wall time to test speed and psutil to test memory usage.